iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp

```diff
@@ -0,0 +1,168 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Conversion/CodegenUtils/MarkerUtils.h"
+#include "iree/compiler/Conversion/CodegenUtils/MatmulCodegenStrategy.h"
+#include "iree/compiler/Conversion/LinalgToLLVM/KernelDispatch.h"
+#include "mlir/Conversion/StandardToSPIRV/ConvertStandardToSPIRV.h"
+#include "mlir/Dialect/Linalg/Transforms/Hoisting.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-linalg-to-llvm-tile-and-vectorize"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+template <typename LinalgOpTy>
+struct TileWorkgroups : public linalg::LinalgBaseTilingPattern {
+  using Base = linalg::LinalgBaseTilingPattern;
+  TileWorkgroups(MLIRContext *context, linalg::LinalgTilingOptions options,
+                 linalg::LinalgMarker marker, PatternBenefit benefit = 1)
+      : Base(LinalgOpTy::getOperationName(), context, options, marker,
+             benefit) {}
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<Value, 4> tensorResults;
+    if (failed(Base::matchAndRewriteBase(op, rewriter, tensorResults)) ||
+        !tensorResults.empty()) {
+      return failure();
+    }
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+}  // namespace
+
+namespace {
+struct TileAndVectorizeWorkgroups
+    : public PassWrapper<TileAndVectorizeWorkgroups, FunctionPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect, AffineDialect, scf::SCFDialect,
+                    vector::VectorDialect>();
+  }
+  void runOnFunction() override;
+};
+}  // namespace
+
+void TileAndVectorizeWorkgroups::runOnFunction() {
+  auto funcOp = getOperation();
+  MLIRContext *context = &getContext();
+
+  // Workgroup tiling.
+  {
+    // First level of tiling patterns. (workgroups memory)
+    OwningRewritePatternList l1patterns, l2patterns;
+    CPUKernelDispatch cpuKernelDispatch;
+    l1patterns.insert<TileWorkgroups<linalg::MatmulOp>,
+                      TileWorkgroups<linalg::BatchMatmulOp>>(
+        context,
+        linalg::LinalgTilingOptions().setTileSizeComputationFunction(
+            [&cpuKernelDispatch](OpBuilder &builder, Operation *operation)
+                -> SmallVector<Value, 4> {
+              return TileSizeFn::get<TilingLevel::Level1Tiles>(
+                  cpuKernelDispatch, builder, operation);
+            }),
+        linalg::LinalgMarker(
+            Identifier::get(getWorkgroupMarker(), context),
+            Identifier::get(getWorkgroupL1TileMarker(), context)));
+
```

Comment:
Nit: based on previous comment. Split the scope here. l2patterns not used before here.